File: V0.2/linkguardMB/Android/app/src/main/java/com/linkguard/app/ble/BluetoothManager.kt
package com.linkguard.app.ble

import android.annotation.SuppressLint
import android.bluetooth.*
import android.bluetooth.le.*
import android.content.Context
import android.os.ParcelUuid
import com.linkguard.app.model.FirmwareLoRaCommand
import com.linkguard.app.model.FirmwareStatusResponse
import com.linkguard.app.model.FirmwareVictim
import com.linkguard.app.model.FirmwareTeamNode
import com.linkguard.app.model.FirmwareReinforcement
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import org.json.JSONObject
import java.util.*
import java.util.LinkedList
import java.util.Queue

// === BLE 常數（對應韌體 rescue.ino UUID）===

object LinkGuardBLE {
    val SERVICE_UUID: UUID      = UUID.fromString("4FAFC201-1FB5-459E-8FCC-C5C9C331914B")
    val STATUS_CHAR_UUID: UUID  = UUID.fromString("BEB5483E-36E1-4688-B7F5-EA07361B26A8")
    val COMMAND_CHAR_UUID: UUID = UUID.fromString("BEB5483E-36E1-4688-B7F5-EA07361B26AB")
    val LORA_CMD_CHAR_UUID: UUID = UUID.fromString("BEB5483E-36E1-4688-B7F5-EA07361B26AC")
    val CCCD_UUID: UUID         = UUID.fromString("00002902-0000-1000-8000-00805f9b34fb")
}

// === 已發現的裝置 ===

data class DiscoveredDevice(
    val address: String,
    val name: String,
    val rssi: Int,
    val device: BluetoothDevice
)

// === JSON 解析 ===

object BLEDataParser {
    fun parseStatusJSON(data: ByteArray): FirmwareStatusResponse? {
        return try {
            val json = JSONObject(String(data, Charsets.UTF_8))
            val id = if (json.has("id")) json.getString("id") else null
            val dept = if (json.has("dept")) json.getString("dept") else null
            val bat = json.getInt("bat")
            val vbat = if (json.has("vbat")) json.getDouble("vbat") else null
            val lvl = json.getInt("lvl")
            val pair = if (json.has("pair")) json.getString("pair") else null
            val victimsArray = json.getJSONArray("victims")
            val victims = mutableListOf<FirmwareVictim>()
            for (i in 0 until victimsArray.length()) {
                val v = victimsArray.getJSONObject(i)
                victims.add(
                    FirmwareVictim(
                        id = v.getString("id"),
                        hr = v.getInt("hr"),
                        bat = v.getInt("bat"),
                        rssi = v.getDouble("rssi"),
                        dist = if (v.has("dist")) v.getString("dist") else null,
                        snr = if (v.has("snr")) v.getDouble("snr") else null,
                        sos = v.getBoolean("sos"),
                        online = v.getBoolean("online")
                    )
                )
            }

            // 團隊節點
            val team = if (json.has("team")) {
                val teamArray = json.getJSONArray("team")
                (0 until teamArray.length()).map { i ->
                    val t = teamArray.getJSONObject(i)
                    FirmwareTeamNode(
                        id = t.getString("id"),
                        dept = t.getString("dept"),
                        bat = t.getInt("bat"),
                        rssi = t.getDouble("rssi"),
                        vc = t.getInt("vc"),
                        online = t.getBoolean("online")
                    )
                }
            } else null

            // 增援請求
            val rf = if (json.has("rf")) {
                val rfArray = json.getJSONArray("rf")
                (0 until rfArray.length()).map { i ->
                    val r = rfArray.getJSONObject(i)
                    FirmwareReinforcement(
                        from = r.getString("from"),
                        msg = r.getString("msg"),
                        loc = r.getString("loc"),
                        ago = r.getInt("ago")
                    )
                }
            } else null

            FirmwareStatusResponse(id, dept, bat, vbat, lvl, pair, victims, team, rf)
        } catch (e: Exception) {
            null
        }
    }

    fun encodeDeptCommand(dept: String): ByteArray = "setdept:$dept".toByteArray(Charsets.UTF_8)
    fun encodeLevelCommand(level: Int): ByteArray = "setlvl:$level".toByteArray(Charsets.UTF_8)
    fun encodePairCommand(pair: String): ByteArray = "setpair:$pair".toByteArray(Charsets.UTF_8)
    fun encodeReinforcementCommand(message: String, location: String): ByteArray =
        "reinforce:$message|$location".toByteArray(Charsets.UTF_8)
    fun encodeReinforcementReply(team: String, accept: Boolean): ByteArray =
        "rf_reply:$team|${if (accept) "JOIN" else "NAK"}".toByteArray(Charsets.UTF_8)
    fun encodeTeamPing(): ByteArray = "team_ping".toByteArray(Charsets.UTF_8)
    fun encodeCommandAck(commandID: String): ByteArray = "cmd_ack:$commandID".toByteArray(Charsets.UTF_8)

    fun parseLoRaCommand(data: ByteArray): FirmwareLoRaCommand? {
        return try {
            val json = JSONObject(String(data, Charsets.UTF_8))
            FirmwareLoRaCommand(
                cmd_id = json.getString("cmd_id"),
                type = json.getString("type"),
                pri = json.optInt("pri", 0),
                title = json.getString("title"),
                detail = json.getString("detail"),
                sender = json.getString("sender")
            )
        } catch (e: Exception) {
            null
        }
    }
}

// === 藍牙管理器 ===

@SuppressLint("MissingPermission")
class BluetoothManager(private val context: Context) {

    private val bluetoothAdapter: BluetoothAdapter? =
        (context.getSystemService(Context.BLUETOOTH_SERVICE) as android.bluetooth.BluetoothManager).adapter

    private var scanner: BluetoothLeScanner? = null
    private var gatt: BluetoothGatt? = null
    private var commandCharacteristic: BluetoothGattCharacteristic? = null

    private val _isScanning = MutableStateFlow(false)
    val isScanning: StateFlow<Boolean> = _isScanning

    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected

    private val _isPoweredOn = MutableStateFlow(false)
    val isPoweredOn: StateFlow<Boolean> = _isPoweredOn

    private val _discoveredDevices = MutableStateFlow<List<DiscoveredDevice>>(emptyList())
    val discoveredDevices: StateFlow<List<DiscoveredDevice>> = _discoveredDevices

    private val _connectedDeviceName = MutableStateFlow<String?>(null)
    val connectedDeviceName: StateFlow<String?> = _connectedDeviceName

    var onStatusUpdate: ((FirmwareStatusResponse) -> Unit)? = null
    var onLoRaCommand: ((String) -> Unit)? = null

    // BLE JSON 累積緩衝區（處理 MTU 截斷）
    private var statusBuffer = ByteArray(0)
    private var statusBufferTimestamp = 0L

    // 排隊 descriptor 寫入（Android 一次只能一個 BLE 操作）
    private val descriptorQueue: Queue<BluetoothGattDescriptor> = LinkedList()
    private var isWritingDescriptor = false

    init {
        _isPoweredOn.value = bluetoothAdapter?.isEnabled == true
    }

    // MARK: 掃描

    fun startScanning() {
        if (bluetoothAdapter?.isEnabled != true) return
        scanner = bluetoothAdapter.bluetoothLeScanner ?: return
        _isScanning.value = true
        _discoveredDevices.value = emptyList()

        val filters = listOf(
            ScanFilter.Builder()
                .setServiceUuid(ParcelUuid(LinkGuardBLE.SERVICE_UUID))
                .build()
        )
        val settings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_POWER)
            .build()

        scanner?.startScan(filters, settings, scanCallback)
    }

    fun stopScanning() {
        scanner?.stopScan(scanCallback)
        _isScanning.value = false
    }

    private val scanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            val device = result.device
            val address = device.address
            val current = _discoveredDevices.value
            if (current.none { it.address == address }) {
                _discoveredDevices.value = current + DiscoveredDevice(
                    address = address,
                    name = device.name ?: "未知裝置",
                    rssi = result.rssi,
                    device = device
                )
            }
        }
    }

    // MARK: 連線

    fun connect(device: DiscoveredDevice) {
        stopScanning()
        val newGatt = device.device.connectGatt(context, false, gattCallback, BluetoothDevice.TRANSPORT_LE)
        if (newGatt == null) {
            android.util.Log.e("BLE", "connectGatt returned null for ${device.device.address}")
            _isConnected.value = false
            return
        }
        gatt = newGatt
    }

    fun disconnect() {
        gatt?.disconnect()
        gatt?.close()
        gatt = null
        _isConnected.value = false
        _connectedDeviceName.value = null
        commandCharacteristic = null
    }

    // MARK: 發送指令

    fun sendDeptChange(dept: String) {
        val char = commandCharacteristic ?: return
        val g = gatt ?: return
        char.value = BLEDataParser.encodeDeptCommand(dept)
        g.writeCharacteristic(char)
    }

    fun sendLevelChange(level: Int) {
        val char = commandCharacteristic ?: return
        val g = gatt ?: return
        char.value = BLEDataParser.encodeLevelCommand(level)
        g.writeCharacteristic(char)
    }

    fun sendPairChange(pair: String) {
        val char = commandCharacteristic ?: return
        val g = gatt ?: return
        char.value = BLEDataParser.encodePairCommand(pair)
        g.writeCharacteristic(char)
    }

    fun sendReinforcement(message: String, location: String) {
        val char = commandCharacteristic ?: return
        val g = gatt ?: return
        char.value = BLEDataParser.encodeReinforcementCommand(message, location)
        g.writeCharacteristic(char)
    }

    fun sendReinforcementReply(team: String, accept: Boolean) {
        val char = commandCharacteristic ?: return
        val g = gatt ?: return
        char.value = BLEDataParser.encodeReinforcementReply(team, accept)
        g.writeCharacteristic(char)
    }

    fun sendTeamPing() {
        val char = commandCharacteristic ?: return
        val g = gatt ?: return
        char.value = BLEDataParser.encodeTeamPing()
        g.writeCharacteristic(char)
    }

    fun sendCommandAck(commandID: String) {
        val char = commandCharacteristic ?: return
        val g = gatt ?: return
        char.value = BLEDataParser.encodeCommandAck(commandID)
        g.writeCharacteristic(char)
    }

    // MARK: GATT 回呼

    private val gattCallback = object : BluetoothGattCallback() {

        override fun onConnectionStateChange(g: BluetoothGatt, status: Int, newState: Int) {
            when (newState) {
                BluetoothProfile.STATE_CONNECTED -> {
                    _isConnected.value = true
                    _connectedDeviceName.value = g.device.name ?: "LinkGuard"
                    // 先請求大 MTU 再探索服務
                    g.requestMtu(512)
                }
                BluetoothProfile.STATE_DISCONNECTED -> {
                    _isConnected.value = false
                    _connectedDeviceName.value = null
                    commandCharacteristic = null
                    statusBuffer = ByteArray(0)
                    gatt?.close()
                    gatt = null
                }
            }
        }

        override fun onMtuChanged(g: BluetoothGatt, mtu: Int, status: Int) {
            android.util.Log.d("BLE", "MTU negotiated: $mtu (status=$status)")
            g.discoverServices()
        }

        override fun onServicesDiscovered(g: BluetoothGatt, status: Int) {
            if (status != BluetoothGatt.GATT_SUCCESS) return
            val service = g.getService(LinkGuardBLE.SERVICE_UUID) ?: return

            // 訂閱狀態 Notify（排隊 descriptor 寫入）
            service.getCharacteristic(LinkGuardBLE.STATUS_CHAR_UUID)?.let { statusChar ->
                g.setCharacteristicNotification(statusChar, true)
                statusChar.getDescriptor(LinkGuardBLE.CCCD_UUID)?.let { desc ->
                    desc.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                    descriptorQueue.add(desc)
                }
            }

            // 保存 Command 特徵
            commandCharacteristic = service.getCharacteristic(LinkGuardBLE.COMMAND_CHAR_UUID)

            // 訂閱 LoRa Command Notify（排隊 descriptor 寫入）
            service.getCharacteristic(LinkGuardBLE.LORA_CMD_CHAR_UUID)?.let { loraChar ->
                g.setCharacteristicNotification(loraChar, true)
                loraChar.getDescriptor(LinkGuardBLE.CCCD_UUID)?.let { desc ->
                    desc.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                    descriptorQueue.add(desc)
                }
            }

            // 開始寫入 descriptor 佇列
            writeNextDescriptor(g)
        }

        override fun onDescriptorWrite(g: BluetoothGatt, descriptor: BluetoothGattDescriptor, status: Int) {
            android.util.Log.d("BLE", "Descriptor write status=$status for ${descriptor.characteristic.uuid}")
            isWritingDescriptor = false
            writeNextDescriptor(g)
        }

        override fun onCharacteristicChanged(
            g: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic
        ) {
            if (characteristic.uuid == LinkGuardBLE.STATUS_CHAR_UUID) {
                val data = characteristic.value ?: return
                android.util.Log.d("BLE", "Status notify: ${data.size} bytes")

                // 累積 BLE 資料（處理 MTU 截斷導致 JSON 不完整的情況）
                // 韌體每次 notify 都是完整 JSON，但可能被 BLE 截斷
                // 檢測策略：如果 data 以 '{' 開頭，清空舊緩衝重新累積
                if (data.isNotEmpty() && data[0] == '{'.code.toByte()) {
                    statusBuffer = data
                    statusBufferTimestamp = System.currentTimeMillis()
                } else {
                    statusBuffer = statusBuffer + data
                }

                // 嘗試解析完整 JSON
                val parsed = BLEDataParser.parseStatusJSON(statusBuffer)
                if (parsed != null) {
                    statusBuffer = ByteArray(0)
                    onStatusUpdate?.invoke(parsed)
                } else if (statusBuffer.size > 4096 ||
                    System.currentTimeMillis() - statusBufferTimestamp > 3000) {
                    // 緩衝區過大或超時，放棄
                    android.util.Log.w("BLE", "Status buffer overflow/timeout (${statusBuffer.size} bytes), clearing")
                    statusBuffer = ByteArray(0)
                }
            } else if (characteristic.uuid == LinkGuardBLE.LORA_CMD_CHAR_UUID) {
                val data = characteristic.value ?: return
                onLoRaCommand?.invoke(String(data, Charsets.UTF_8))
            }
        }
    }

    private fun writeNextDescriptor(g: BluetoothGatt) {
        if (isWritingDescriptor) return
        val desc = descriptorQueue.poll() ?: return
        isWritingDescriptor = true
        g.writeDescriptor(desc)
    }
}

File: v0.4_Rebuild_INSARAG/apps/ios/LinkGuard-EMT/LinkGuard-EMT/V02FieldApp/BluetoothManager.swift
import Foundation
import CoreBluetooth
import Combine

// MARK: - BLE 常數（對應 Heltec WiFi LoRa 32 V3 韌體預留 BLE UUID）

struct LinkGuardBLE {
    // 主服務
    static let serviceUUID      = CBUUID(string: "4FAFC201-1FB5-459E-8FCC-C5C9C331914B")
    // 狀態特徵：韌體端送出與 /up 相同的 JSON（bat, lvl, victims[]）
    static let statusUUID       = CBUUID(string: "BEB5483E-36E1-4688-B7F5-EA07361B26A8")
    // 指令特徵：App 端寫入控制指令（setdept, setlvl）
    static let commandUUID      = CBUUID(string: "BEB5483E-36E1-4688-B7F5-EA07361B26AB")
    // LoRa 命令特徵：韌體端轉發來自指揮中心的 LoRa 命令（Notify）
    static let loraCommandUUID  = CBUUID(string: "BEB5483E-36E1-4688-B7F5-EA07361B26AC")
}

// MARK: - 已發現裝置

struct DiscoveredDevice: Identifiable {
    let id: UUID
    let name: String
    let rssi: Int
    let peripheral: CBPeripheral
}

// MARK: - BLE 資料解析

struct BLEDataParser {

    /// 解析韌體 JSON（同 /up 端點格式）
    static func parseStatusJSON(_ data: Data) -> FirmwareStatusResponse? {
        do {
            return try JSONDecoder().decode(FirmwareStatusResponse.self, from: data)
        } catch {
            let preview = String(data: data.prefix(200), encoding: .utf8) ?? "<binary>"
            print("[BLE] JSON parse FAIL (\(data.count) bytes): \(error)")
            print("[BLE] Raw data: \(preview)")
            return nil
        }
    }

    /// 編碼 setdept 指令
    static func encodeDeptCommand(_ dept: String) -> Data? {
        "setdept:\(dept)".data(using: .utf8)
    }

    /// 編碼 setlvl 指令
    static func encodeLevelCommand(_ level: Int) -> Data? {
        "setlvl:\(level)".data(using: .utf8)
    }

    /// 編碼 setpair 指令
    static func encodePairCommand(_ pair: String) -> Data? {
        "setpair:\(pair)".data(using: .utf8)
    }

    /// 編碼命令回執（通知韌體 App 已收到命令）
    static func encodeCommandAck(_ commandID: String) -> Data? {
        "cmd_ack:\(commandID)".data(using: .utf8)
    }

    /// 解析韌體轉發的 LoRa 命令 JSON
    /// 格式：{"cmd_id":"xxx","type":"evacuation","pri":2,"title":"立即撤離","detail":"...","sender":"HQ-Alpha"}
    static func parseLoRaCommand(_ data: Data) -> FirmwareLoRaCommand? {
        do {
            return try JSONDecoder().decode(FirmwareLoRaCommand.self, from: data)
        } catch {
            print("[BLE] LoRa command decode failed: \(error.localizedDescription)")
            return nil
        }
    }
}

/// 韌體轉發的 LoRa 指揮命令格式
struct FirmwareLoRaCommand: Decodable {
    let cmd_id: String          // 命令唯一 ID
    let type: String            // 命令類型 key（對應 CommandType.rawValue 或英文 key）
    let pri: Int                // 優先等級 0=routine, 1=urgent, 2=critical
    let title: String           // 命令標題
    let detail: String          // 命令詳情
    let sender: String          // 發送者代號

    /// 轉換為 App 內部的 CommandOrder
    func toCommandOrder() -> CommandOrder {
        let priority = CommandPriority(rawValue: pri) ?? .routine
        let commandType = CommandType.fromKey(type) ?? .statusReport

        return CommandOrder(
            id: UUID(uuidString: cmd_id) ?? UUID(),
            type: commandType,
            priority: priority,
            title: title,
            detail: detail,
            sender: sender,
            time: Date(),
            isRead: false
        )
    }
}

// MARK: - 藍牙管理器

class BluetoothManager: NSObject, ObservableObject {
    @Published var isScanning = false
    @Published var isConnected = false
    @Published var isPoweredOn = false
    @Published var discoveredDevices: [DiscoveredDevice] = []
    @Published var connectedDeviceName: String?

    private var centralManager: CBCentralManager!
    private var connectedPeripheral: CBPeripheral?
    private var commandCharacteristic: CBCharacteristic?
    private var statusCharacteristic: CBCharacteristic?

    /// BLE JSON 累積緩衝區（處理 MTU 截斷）
    private var statusBuffer = Data()
    private var statusBufferTimestamp = Date()
    private var reconnectTimer: Timer?
    private var reconnectDelay: TimeInterval = 2

    /// 韌體狀態更新回呼
    var onStatusUpdate: ((FirmwareStatusResponse) -> Void)?
    /// LoRa 指揮命令回呼
    var onLoRaCommand: ((FirmwareLoRaCommand) -> Void)?

    override init() {
        super.init()
        centralManager = CBCentralManager(delegate: self, queue: nil)
    }

    deinit {
        reconnectTimer?.invalidate()
        reconnectTimer = nil
    }

    // MARK: 掃描

    func startScanning() {
        guard centralManager.state == .poweredOn else { return }
        isScanning = true
        discoveredDevices.removeAll()
        centralManager.scanForPeripherals(
            withServices: [LinkGuardBLE.serviceUUID],
            options: [CBCentralManagerScanOptionAllowDuplicatesKey: false]
        )
    }

    func stopScanning() {
        centralManager.stopScan()
        isScanning = false
    }

    // MARK: 連線

    func connect(to device: DiscoveredDevice) {
        stopScanning()
        connectedPeripheral = device.peripheral
        centralManager.connect(device.peripheral, options: nil)
    }

    func disconnect() {
        reconnectTimer?.invalidate()
        reconnectDelay = 2
        guard let p = connectedPeripheral else { return }
        connectedPeripheral = nil
        connectedDeviceName = nil
        centralManager.cancelPeripheralConnection(p)
    }

    // MARK: 發送指令

    func sendDeptChange(_ dept: String) {
        guard let char = commandCharacteristic,
              let p = connectedPeripheral,
              let data = BLEDataParser.encodeDeptCommand(dept)
        else { return }
        p.writeValue(data, for: char, type: .withResponse)
    }

    func sendLevelChange(_ level: Int) {
        guard let char = commandCharacteristic,
              let p = connectedPeripheral,
              let data = BLEDataParser.encodeLevelCommand(level)
        else { return }
        p.writeValue(data, for: char, type: .withResponse)
    }

    func sendPairChange(_ pair: String) {
        guard let char = commandCharacteristic,
              let p = connectedPeripheral,
              let data = BLEDataParser.encodePairCommand(pair)
        else { return }
        p.writeValue(data, for: char, type: .withResponse)
    }

    func sendReinforcement(message: String, location: String) {
        guard let char = commandCharacteristic,
              let p = connectedPeripheral,
              let data = "reinforce:\(message)|\(location)".data(using: .utf8)
        else { return }
        p.writeValue(data, for: char, type: .withResponse)
    }

    func sendReinforcementReply(to team: String, accept: Bool) {
        guard let char = commandCharacteristic,
              let p = connectedPeripheral,
              let data = "rf_reply:\(team)|\(accept ? "JOIN" : "NAK")".data(using: .utf8)
        else { return }
        p.writeValue(data, for: char, type: .withResponse)
    }

    func sendTeamPing() {
        guard let char = commandCharacteristic,
              let p = connectedPeripheral,
              let data = "team_ping".data(using: .utf8)
        else { return }
        p.writeValue(data, for: char, type: .withResponse)
    }

    /// 發送命令回執（通知韌體 App 已收到 LoRa 命令）
    func sendCommandAck(_ commandID: String) {
        guard let char = commandCharacteristic,
              let p = connectedPeripheral,
              let data = BLEDataParser.encodeCommandAck(commandID)
        else { return }
        p.writeValue(data, for: char, type: .withResponse)
    }
}

// MARK: - CBCentralManagerDelegate

extension BluetoothManager: CBCentralManagerDelegate {

    func centralManagerDidUpdateState(_ central: CBCentralManager) {
        isPoweredOn = central.state == .poweredOn
        if central.state != .poweredOn {
            isConnected = false
            isScanning = false
        }
    }

    func centralManager(_ central: CBCentralManager,
                         didDiscover peripheral: CBPeripheral,
                         advertisementData: [String: Any],
                         rssi RSSI: NSNumber) {
        guard !discoveredDevices.contains(where: { $0.id == peripheral.identifier }) else { return }
        let name = peripheral.name ?? L("未知裝置")
        discoveredDevices.append(
            DiscoveredDevice(id: peripheral.identifier, name: name, rssi: RSSI.intValue, peripheral: peripheral)
        )
    }

    func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
        isConnected = true
        reconnectTimer?.invalidate()
        reconnectDelay = 2  // 重連成功，重設退避計時
        connectedDeviceName = peripheral.name ?? "LinkGuard"
        peripheral.delegate = self
        print("[BLE] Connected to \(peripheral.name ?? "?"), discovering services...")
        peripheral.discoverServices([LinkGuardBLE.serviceUUID])
    }

    func centralManager(_ central: CBCentralManager,
                         didDisconnectPeripheral peripheral: CBPeripheral,
                         error: Error?) {
        isConnected = false
        commandCharacteristic = nil
        statusCharacteristic = nil
        statusBuffer.removeAll()
        // 自動重連（指數退避：2s → 4s → 8s … 最多 60s）
        if connectedPeripheral != nil {
            print("[BLE] Disconnected, will retry in \(reconnectDelay)s...")
            reconnectTimer?.invalidate()
            reconnectTimer = Timer.scheduledTimer(withTimeInterval: reconnectDelay, repeats: false) { [weak self] _ in
                guard let self, let p = self.connectedPeripheral, self.centralManager.state == .poweredOn else { return }
                print("[BLE] Attempting reconnect...")
                self.centralManager.connect(p, options: nil)
            }
            reconnectDelay = min(reconnectDelay * 2, 60)
        }
    }

    func centralManager(_ central: CBCentralManager,
                         didFailToConnect peripheral: CBPeripheral,
                         error: Error?) {
        isConnected = false
        connectedPeripheral = nil
    }
}

// MARK: - CBPeripheralDelegate

extension BluetoothManager: CBPeripheralDelegate {

    func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
        if let error = error {
            print("[BLE] Service discovery error: \(error)")
            return
        }
        guard let services = peripheral.services else {
            print("[BLE] No services found")
            return
        }
        print("[BLE] Found \(services.count) services")
        for service in services {
            peripheral.discoverCharacteristics([
                LinkGuardBLE.statusUUID,
                LinkGuardBLE.commandUUID,
                LinkGuardBLE.loraCommandUUID
            ], for: service)
        }
    }

    func peripheral(_ peripheral: CBPeripheral,
                     didDiscoverCharacteristicsFor service: CBService,
                     error: Error?) {
        guard let chars = service.characteristics else {
            print("[BLE] No characteristics found")
            return
        }
        print("[BLE] Found \(chars.count) characteristics")
        for char in chars {
            switch char.uuid {
            case LinkGuardBLE.statusUUID:
                statusCharacteristic = char
                peripheral.setNotifyValue(true, for: char)
                print("[BLE] Subscribed to status notifications")
            case LinkGuardBLE.commandUUID:
                commandCharacteristic = char
                print("[BLE] Found command characteristic")
            case LinkGuardBLE.loraCommandUUID:
                peripheral.setNotifyValue(true, for: char)
                print("[BLE] Subscribed to LoRa command notifications")
            default:
                break
            }
        }
    }

    func peripheral(_ peripheral: CBPeripheral,
                     didUpdateValueFor characteristic: CBCharacteristic,
                     error: Error?) {
        if let error = error {
            print("[BLE] Read error: \(error)")
            return
        }
        guard let data = characteristic.value, !data.isEmpty else {
            print("[BLE] Empty data for \(characteristic.uuid)")
            return
        }

        print("[BLE] Received \(data.count) bytes from \(characteristic.uuid.uuidString.prefix(8))")

        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            switch characteristic.uuid {
            case LinkGuardBLE.statusUUID:
                // 累積 BLE 資料（處理 MTU 截斷導致 JSON 不完整的情況）
                // 韌體每次 notify 都是完整 JSON，但可能被 BLE 截斷
                // 檢測策略：如果 data 以 '{' 開頭，清空舊緩衝重新累積
                if let firstByte = data.first, firstByte == UInt8(ascii: "{") {
                    self.statusBuffer = data
                    self.statusBufferTimestamp = Date()
                } else {
                    self.statusBuffer.append(data)
                }

                // 嘗試解析完整 JSON
                if let status = BLEDataParser.parseStatusJSON(self.statusBuffer) {
                    self.statusBuffer.removeAll()
                    self.onStatusUpdate?(status)
                } else if self.statusBuffer.count > 4096 ||
                          Date().timeIntervalSince(self.statusBufferTimestamp) > 3 {
                    // 緩衝區過大或超時，放棄
                    print("[BLE] Status buffer overflow/timeout (\(self.statusBuffer.count) bytes), clearing")
                    self.statusBuffer.removeAll()
                }
            case LinkGuardBLE.loraCommandUUID:
                if let command = BLEDataParser.parseLoRaCommand(data) {
                    self.onLoRaCommand?(command)
                }
            default:
                break
            }
        }
    }

    // MARK: Notify 訂閱結果

    func peripheral(_ peripheral: CBPeripheral,
                     didUpdateNotificationStateFor characteristic: CBCharacteristic,
                     error: Error?) {
        if let error = error {
            print("[BLE] ❌ Notify subscribe FAILED for \(characteristic.uuid.uuidString.prefix(8)): \(error)")
        } else {
            print("[BLE] ✅ Notify \(characteristic.isNotifying ? "ON" : "OFF") for \(characteristic.uuid.uuidString.prefix(8))")
        }
    }

    // MARK: 服務變更

    func peripheral(_ peripheral: CBPeripheral, didModifyServices invalidatedServices: [CBService]) {
        print("[BLE] Services invalidated: \(invalidatedServices.map { $0.uuid.uuidString })")
        // 重新探索服務
        peripheral.discoverServices([LinkGuardBLE.serviceUUID])
    }

    // MARK: Write 結果

    func peripheral(_ peripheral: CBPeripheral,
                     didWriteValueFor characteristic: CBCharacteristic,
                     error: Error?) {
        if let error = error {
            print("[BLE] ❌ Write FAILED: \(error)")
        } else {
            print("[BLE] ✅ Write OK to \(characteristic.uuid.uuidString.prefix(8))")
        }
    }
}
